### PR TITLE
Add method to retrieve federated session info

### PR DESCRIPTION
**Pull request by ThaminduR**
### Proposed changes in this pull request

Add `getSessionDetails` for retrieving federated session info during the federated idp initiated logout

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java [hunk 1]
```diff
@@ -89,4 +98,31 @@ public UserSession getSession(String sessionId) throws SessionManagementServerEx
...
         return null;
     }
+
+    /**
+     *
+     * @param fedIdpSessionId
+     * @return A map containing federated authentication session details
+     * @throws SessionManagementServerException
+     */
+    public Map<String, String> getSessionDetails(String fedIdpSessionId) throws SessionManagementServerException {
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java:108]: Do we need a separate method here?
What happens if we use the getSession() method with the federatedSessionID?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java [hunk 2]
```diff
@@ -22,11 +22,14 @@
...
+import java.util.Map;
+
 /**
  * Perform operations for {@link UserSession}.
  */
 public interface UserSessionDAO {
 
     UserSession getSession(String sessionId) throws SessionManagementServerException;
 
+    Map<String, String> getSessionDetails(String oidcSId) throws SessionManagementServerException;
```
[the remark this answers is not in the document]
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java]: Fixed in 5a59267

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java [hunk 3]
```diff
@@ -120,6 +120,9 @@
     public static final String SQL_STORE_FEDERATED_AUTH_SESSION_INFO = "INSERT INTO IDN_FED_AUTH_SESSION_MAPPING "
             + "(IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, PROTOCOL_TYPE) VALUES (?, ?, ?, ?, ?)";
 
+    public static final String SQL_GET_FEDERATED_AUTH_SESSION_INFO_BY_SESSION_ID = "SELECT * FROM" +
```
[the remark this answers is not in the document]
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java]: Changed in 5a59267

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java (hunk 1 above)
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java:109]: We can't use the getSession() to retrieve the federated authentication session info.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java [hunk 4]
```diff
@@ -22,11 +22,21 @@
...
     UserSession getSession(String sessionId) throws SessionManagementServerException;
 
+    /**
+     * Get federated authentication session mapping info during the federated idp init logout
+     *
+     * @param fedIdpSessionId
+     * @return A map containing federated authentication session details
+     * @throws SessionManagementServerException
+     */
+    Map<String, String> getSessionDetails(String fedIdpSessionId) throws SessionManagementServerException;
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java:41]: Shall we rename the method to reflect that the session retrieved is for a federated user?

Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java]: Renamed in 845675d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java [hunk 5]
```diff
@@ -89,4 +95,36 @@ public UserSession getSession(String sessionId) throws SessionManagementServerEx
...
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false);
+             PreparedStatement preparedStatement =
+                     connection.prepareStatement(SQLQueries.SQL_GET_FEDERATED_AUTH_SESSION_INFO_BY_SESSION_ID)) {
+            preparedStatement.setString(1, fedIdpSessionId);
+            try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                FederatedUserSession federatedUserSession = new FederatedUserSession();
+                if (resultSet.next()) {
+                    federatedUserSession.setIdPSessionId(fedIdpSessionId);
+                    federatedUserSession.setSessionId(resultSet.getString("SESSION_ID"));
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java:117]: Let's use constants for column names.

Also can you take a look at UserSession getSession(String sessionId) and see if we can follow the same approach with JdbcTemplates

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.application.authentication.framework.model;
+
+/**
+ * FederatedUserSession is the class that represents a session object of a federated user
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java:22]:
```suggestion
 * Represents a session of a federated user
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java [hunk 7]
```diff
@@ -120,6 +120,10 @@
     public static final String SQL_STORE_FEDERATED_AUTH_SESSION_INFO = "INSERT INTO IDN_FED_AUTH_SESSION_MAPPING "
             + "(IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, PROTOCOL_TYPE) VALUES (?, ?, ?, ?, ?)";
 
+    // Get federated authentication session details using the IDP session id
+    public static final String SQL_GET_FEDERATED_AUTH_SESSION_INFO_BY_SESSION_ID = "SELECT SESSION_ID, IDP_NAME FROM " +
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java:124]: Query needs to modify to return all the required info for the FederatedUserSession.

Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java]: Fixed in 4ac6106

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java (hunk 5 above)
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java]: Implemented with the jdbc templates in 4ac6106

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * IdPSessionId
+ * SessionId
+ * IdPName
+ * AuthenticatorName
+ * ProtocolType
+ */
+public class FederatedUserSession {
+
+    private String IdPSessionId;
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java:32]: variable names should start with simple letters

**Comment by ThaminduR** [comment]
New PR - https://github.com/wso2/carbon-identity-framework/pull/3340

**Pull request closed without merging**